libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/Cddl.hs

```diff
@@ -0,0 +1,238 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.Binary.Cddl where
+
+import Cardano.Ledger.Binary
+import Control.Monad (forM_)
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.ByteString.Lazy.Char8 as BSL8
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import Data.Typeable
+import System.IO (hClose)
+import System.Process.Typed (
+  ExitCode (..),
+  ProcessConfig,
+  byteStringInput,
+  proc,
+  readProcess,
+  setStdin,
+ )
+import Test.Cardano.Ledger.Binary.RoundTrip
+import Test.Hspec
+import UnliftIO.Temporary (withTempFile)
+
+newtype Cddl = Cddl {unCddl :: BSL.ByteString}
+
+newtype Cbor = Cbor {unCbor :: BSL.ByteString}
+
+newtype DiagCbor = DiagCbor {unDiagCbor :: BSL.ByteString}
+
+data CddlVarFile = CddlVarFile
+  { cddlVarFilePath :: !FilePath
+  -- ^ File that contains the Cddl data included in this type
+  , cddlVarName :: !T.Text
+  -- ^ Name of the variable being tested
+  , cddlVarData :: !Cddl
+  -- ^ Full CDDL spec with @output=varName@ prefix
+  , cddlVarDiagCbor :: ![DiagCbor]
+  -- ^ Generated CBOR data from the above CDDL spec
+  }
+
+data CddlData = CddlData
+  { cddlData :: !Cddl
+  , cddlNumExamples :: !Int
+  }
+
+beforeAllCddlFile :: HasCallStack => Int -> IO [BSL.ByteString] -> SpecWith CddlData -> Spec
```

Comment:
You need to be familiar with `hspec` to know it.
It is actually the reason, why this approach is faster, because it prepares the state for all of the tests in its scope only once. As far as I know `tasty` simply does not have capability like that.

I'll add some haddock to `beforeAllCddlFile`